### PR TITLE
New references for Dimension Scale APIs

### PR DESCRIPTION
**Pull request by jhendersonHDF**
Work by @epourmal to enable use of new-style references in the dimension scale APIs.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1053,6 +1053,16 @@ if (EXISTS "${HDF5_SOURCE_DIR}/hl" AND IS_DIRECTORY "${HDF5_SOURCE_DIR}/hl")
   endif ()
 endif ()
 
+#-----------------------------------------------------------------------------
+# Option to use new-style references with dimension scale APIs
+#-----------------------------------------------------------------------------
+if (HDF5_BUILD_HL_LIB)
+  option (HDF5_DIMENSION_SCALES_NEW_REF  "Use new-style references with dimension scale APIs" OFF)
```
Comment (byrnHDF) [inline CMakeLists.txt]: This option should be in the HL CMakeLists.txt file as it only applies to that library.

File: config/cmake/H5pubconf.h.in [hunk 2]
```diff
@@ -406,6 +406,9 @@
 /* Define if the high-level library headers should be included in hdf5.h */
 #cmakedefine H5_INCLUDE_HL @H5_INCLUDE_HL@
 
+/* Define if new-style references should be used with dimension scales */
```
Comment (byrnHDF) [inline config/cmake/H5pubconf.h.in:409]: I like it next to the HL define - however I believe the defines are supposed to be sorted alphabetically.

File: hl/src/H5DS.c [hunk 3]
```diff
@@ -106,22 +135,34 @@ H5DSset_scale(hid_t dsid, const char *dimname)
...
-    hobj_ref_t  ref_to_ds;    /* reference to the DS */
-    hobj_ref_t  ref_j;        /* iterator reference */
-    hvl_t *     buf = NULL;   /* VL buffer to store in the attribute */
-    hid_t       dsid_j;       /* DS dataset ID in DIMENSION_LIST */
+    int      has_dimlist;
+    int      has_reflist;
+    int      is_ds;
+    hssize_t nelmts;
+    hid_t    sid, sid_w; /* space ID */
+    hid_t    tid  = -1;  /* attribute type ID */
```
Comment (byrnHDF) [inline hl/src/H5DS.c]: If you feel like it, use the H5I_INVALID_HID inside of -1.

File: config/cmake/H5pubconf.h.in (hunk 2 above)
Comment (jhendersonHDF) [reply config/cmake/H5pubconf.h.in:409]: Looking at just the first few lines of H5pubconf.h.in:

#cmakedefine H5_HAVE_WINDOWS @H5_HAVE_WINDOWS@
#cmakedefine H5_HAVE_MINGW @H5_HAVE_MINGW@
#cmakedefine H5_HAVE_WIN32_API @H5_HAVE_WIN32_API@
#cmakedefine H5_HAVE_VISUAL_STUDIO @H5_HAVE_VISUAL_STUDIO@
#cmakedefine H5_DEFAULT_PLUGINDIR "@H5_DEFAULT_PLUGINDIR@"
#cmakedefine H5_DEV_T_IS_SCALAR @H5_DEV_T_IS_SCALAR@

followed by the H5_F category, followed by more H5_HAVE category entries. Seems we've already kind of broken that, though I can move the entry if we want to keep trying to organize it.

Comment (byrnHDF) [reply config/cmake/H5pubconf.h.in:409]: Yeah, I think someone likes them in an order I thought was alpha. Maybe I'm wrong.

File: CMakeLists.txt (hunk 1 above)
Comment (jhendersonHDF) [reply CMakeLists.txt]: Moved

File: hl/src/H5DS.c (hunk 3 above)
Comment (jhendersonHDF) [reply hl/src/H5DS.c]: Fixed

File: hl/src/H5DS.c [hunk 4]
```diff
@@ -18,7 +18,36 @@
 
 /* Local routines */
 static herr_t H5DS_is_reserved(hid_t did);
-static hid_t  H5DS_get_REFLIST_type(void);
+
+/*-------------------------------------------------------------------------
+ * Function: H5DSwith_new_ref
+ *
+ * Purpose: Detremines if new references are used with dimension scales.
```
Comment (soumagne) [inline hl/src/H5DS.c:25]: typo

**Review by soumagne — APPROVED** [review]

**Review by byrnHDF — APPROVED** [review]

**Pull request merged**